### PR TITLE
[basicdataset] Add WikiText-2 dataset

### PR DESCRIPTION
**Pull request by dandansamax**
## Description ##

Add WikiText-2 dataset to basicdatasets.

- We referred to the code of cookStackexchange dataset. Function `WikiText2.getData()` will return the path of original text data.

### CODE REVIEW

**Comment by dandansamax** [comment]
Close #45

**Comment by dandansamax** [comment]
BTW, could you help me upload the metadata to the server and write remote testcases?

File: basicdataset/src/main/java/ai/djl/basicdataset/nlp/WikiText2.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+import java.nio.file.Path;
+
+/**
+ * The WikiText language modeling dataset is a collection of over 100 million tokens extracted from
+ * the set of verified Good and Featured articles on Wikipedia.
+ */
+public class WikiText2 implements RawDataset<Path> {
+
+    private static final String VERSION = "1.0";
+    private static final String ARTIFACT_ID = "WikiText-2";
```
Comment (frankfliu) [inline basicdataset/src/main/java/ai/djl/basicdataset/nlp/WikiText2.java:36]:
```suggestion
    private static final String ARTIFACT_ID = "wikitext-2";
```

File: basicdataset/src/main/java/ai/djl/basicdataset/nlp/WikiText2.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    @Override
+    public void prepare(Progress progress) throws IOException {
+        if (prepared) {
+            return;
+        }
+
+        Artifact artifact = mrl.getDefaultArtifact();
+        mrl.prepare(artifact, progress);
+        Artifact.Item item;
+        item = artifact.getFiles().get("WikiText-2");
```
Comment (frankfliu) [inline basicdataset/src/main/java/ai/djl/basicdataset/nlp/WikiText2.java:73]:
```suggestion
        item = artifact.getFiles().get("wikitext-2");
```

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+  "metadataVersion": "0.2",
+  "resourceType": "dataset",
+  "application": "nlp",
+  "groupId": "ai.djl.basicdataset",
+  "artifactId": "WikiText-2",
```
Comment (frankfliu) [inline basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json:6]:
```suggestion
  "artifactId": "wikitext-2",
```

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  "name": "WikiText-2",
```
Comment (frankfliu) [inline basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json:7]:
```suggestion
  "name": "wikitext-2",
```

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    "license": {
+      "name": "Creative Commons Attribution-ShareAlike License",
+      "url": "https://en.wikipedia.org/wiki/Wikipedia:Text_of_Creative_Commons_Attribution-ShareAlike_3.0_Unported_License"
+    }
+  },
+  "artifacts": [
+    {
+      "version": "1.0",
+      "snapshot": false,
+      "name": "WikiText-2",
```
Comment (frankfliu) [inline basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json:20]:
```suggestion
      "name": "wikitext-2",
```

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      "files": {
+        "WikiText-2": {
+          "uri": "https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip",
+          "sha1Hash": "3c914d17d80b1459be871a5039ac23e752a53cbe",
+          "size": 4475746
```
Comment (frankfliu) [inline basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json:25]:
```suggestion
          "name": "",
          "size": 4475746
```

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (frankfliu) [inline basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json:1]: Please use lower case folder name:

basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/wikitext-2/metadata.json

File: basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import ai.djl.repository.Repository;
+import ai.djl.training.dataset.Dataset;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class WikiText2Test {
+    @Test
```
Comment (frankfliu) [inline basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java:25]:
```suggestion
    
    @Test
    public void testWikiTest2TrainRemote() throws IOException {
        WikiText2 trainingSet = WikiText2.builder().optUsage(Dataset.Usage.TRAIN).build();
        Path path = trainingSet.getData();
        Assert.assertTrue(Files.isRegularFile(path));
    }
    
    @Test
```

**Comment by frankfliu** [comment]
@dandansamax 

Thanks for your contribution. I have uploaded modified version of `metadata.json` file to our repo.
Please update your PR and push it again (I'm not able to push my changes to your repo, enabling `allow committer to update PR option` can make collaboration easier).

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+      "url": "https://en.wikipedia.org/wiki/Wikipedia:Text_of_Creative_Commons_Attribution-ShareAlike_3.0_Unported_License"
+    }
+  },
+  "artifacts": [
+    {
+      "version": "1.0",
+      "snapshot": false,
+      "name": "wikitext-2",
+      "files": {
+        "WikiText-2": {
```
Comment (frankfliu) [inline basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json:22]:
```suggestion
        "wikitext-2": {
```

File: basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class WikiText2Test {
+    
+    @Test
+    public void testWikiTest2TrainRemote() throws IOException {
```
Comment (frankfliu) [inline basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java:27]:
```suggestion
    public void testWikiText2TrainRemote() throws IOException {
```

File: basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import ai.djl.repository.Repository;
+import ai.djl.training.dataset.Dataset;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class WikiText2Test {
+
```
Comment (frankfliu) [inline basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java:25]:
```suggestion

```

File: basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+public class WikiText2Test {
+    
+    @Test
+    public void testWikiTest2TrainRemote() throws IOException {
+        WikiText2 trainingSet = WikiText2.builder().optUsage(Dataset.Usage.TRAIN).build();
+        Path path = trainingSet.getData();
+        Assert.assertTrue(Files.isRegularFile(path));
+    }
+
```
Comment (frankfliu) [inline basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java:32]:
```suggestion

```

File: basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    
+    @Test
+    public void testWikiTest2TrainRemote() throws IOException {
+        WikiText2 trainingSet = WikiText2.builder().optUsage(Dataset.Usage.TRAIN).build();
+        Path path = trainingSet.getData();
+        Assert.assertTrue(Files.isRegularFile(path));
+    }
+    
+    @Test
+    public void testWikiTest2TrainLocal() throws IOException {
```
Comment (frankfliu) [inline basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java:34]:
```suggestion
    public void testWikiText2TrainLocal() throws IOException {
```

File: basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        Repository repository = Repository.newInstance("test", "src/test/resources/mlrepo");
+
+        WikiText2 trainingSet =
+                WikiText2.builder().optRepository(repository).optUsage(Dataset.Usage.TRAIN).build();
+        Path path = trainingSet.getData();
+        Assert.assertTrue(Files.isRegularFile(path));
+    }
+
+    @Test
+    public void testWikiTest2TestLocal() throws IOException {
```
Comment (frankfliu) [inline basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java:44]:
```suggestion
    public void testWikiText2TestLocal() throws IOException {
```

File: basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+        Repository repository = Repository.newInstance("test", "src/test/resources/mlrepo");
+
+        WikiText2 trainingSet =
+                WikiText2.builder().optRepository(repository).optUsage(Dataset.Usage.TEST).build();
+        Path path = trainingSet.getData();
+        Assert.assertTrue(Files.isRegularFile(path));
+    }
+
+    @Test
+    public void testWikiTest2ValidationLocal() throws IOException {
```
Comment (frankfliu) [inline basicdataset/src/test/java/ai/djl/basicdataset/WikiText2Test.java:54]:
```suggestion
    public void testWikiText2ValidationLocal() throws IOException {
```

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json (hunk 7 above)
Comment (frankfliu) [reply basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json:1]: > 

Linux file name is case sensitive. If you are on mac/Windows, you won't be able to rename the folder to lower case. You might need to remove WikiText-2 folder complete and recreate the folder with lower case.

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/wikitext-2/metadata.json [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+      "url": "https://en.wikipedia.org/wiki/Wikipedia:Text_of_Creative_Commons_Attribution-ShareAlike_3.0_Unported_License"
+    }
+  },
+  "artifacts": [
+    {
+      "version": "1.0",
+      "snapshot": false,
+      "name": "wikitext-2",
+      "files": {
+        "WikiText-2": {
```
Comment (frankfliu) [inline basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/wikitext-2/metadata.json:22]:
```suggestion
        "wikitext-2": {
```

**Review by frankfliu — APPROVED** [review]

File: basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json (hunk 7 above)
Comment (dandansamax) [reply basicdataset/src/test/resources/mlrepo/dataset/nlp/ai/djl/basicdataset/WikiText-2/metadata.json]: Thank you, I have changed it.

**Comment by frankfliu** [comment]
@dandansamax 

Please format the java code, there is white space in the test code:
```
./gradlew formatJava
```

And then commit the format changes.

**Comment by dandansamax** [comment]
Thank you so much for your patient help! I will enable edits from maintainers next time.

**Pull request merged**